feat(iotsitewise): add portal, project, gateway, dashboard, asset, asset model and access policy

npellegrin opened the pull request:
Closes #419

npellegrin commented:
New resource types:

  - IoTSiteWiseAsset
  - IoTSiteWiseAssetModel
  - IoTSiteWiseDashboard
  - IoTSiteWiseGateway
  - IoTSiteWisePortal
  - IoTSiteWiseProject
  - IoTSiteWiseAccessPolicy

Scan with feature activated:

```
eu-west-1 - IoTSiteWisePortal - 9446b069-b1ad-4a41-9d77-70b5df677cf2 - [name: "test Portal"] - would remove
eu-west-1 - IoTSiteWiseGateway - 65cb1349-be5d-42b1-ada5-13b2d459e8c9 - [name: "Gateway-B8Enr_b-p"] - would remove
eu-west-1 - IoTSiteWiseProject - 3a0aa9bf-bd59-4289-a86f-2e939cca0086 - [name: "test Project"] - would remove
eu-west-1 - IoTSiteWiseDashboard - bbe95e19-c7bc-447c-9a96-183fd694eef9 - [name: "New dashboard"] - would remove
eu-west-1 - IoTSiteWiseAsset - 06214240-7e09-4528-a002-c8e46f8217a3 - [name: "Oven_1", status: "ACTIVE"] - would remove
eu-west-1 - IoTSiteWiseAsset - c4acd8c3-9f56-4a0f-a44e-e98e4ea19922 - [name: "HorizontalConveyor_1", status: "ACTIVE"] - would remove
eu-west-1 - IoTSiteWiseAsset - 7bdf4401-7276-44f3-ac08-3cb133ba2c3d - [name: "testMaker_1", status: "ACTIVE"] - would remove
eu-west-1 - IoTSiteWiseAsset - 21388a7b-739c-4d27-82d7-dececb2b9560 - [name: "VerticalConveyor_1", status: "ACTIVE"] - would remove
eu-west-1 - IoTSiteWiseAsset - 051548cb-383a-4380-bad4-75086dead17b - [name: "Mixer_1", status: "ACTIVE"] - would remove
eu-west-1 - IoTSiteWiseAssetModel - 41dba598-ec81-465f-b96a-bad704924fe5 - [name: "Oven_Model", status: "ACTIVE"] - would remove
eu-west-1 - IoTSiteWiseAssetModel - 74bd3cd0-5798-4b3a-930c-b0a1e3469492 - [name: "Mixer_Model", status: "ACTIVE"] - would remove
eu-west-1 - IoTSiteWiseAssetModel - a3f6ff1f-4c29-4f45-8c80-b7a2f8b7f236 - [name: "testMaker_Model", status: "ACTIVE"] - would remove
eu-west-1 - IoTSiteWiseAssetModel - b5ea342a-260a-4cb3-9d21-b4a2c4198b8f - [name: "HorizontalConveyor_Model", status: "ACTIVE"] - would remove
eu-west-1 - IoTSiteWiseAssetModel - ba064ef9-1b60-44ed-a5e6-efb98e22172a - [name: "VerticalConveyor_Model", status: "ACTIVE"] - would remove
Scan complete: 14 total, 14 nukeable, 0 filtered.
```

Deletion log:

```
eu-west-1 - IoTSiteWisePortal - 9446b069-b1ad-4a41-9d77-70b5df677cf2 - [name: "test Portal"] - removed
eu-west-1 - IoTSiteWiseProject - 3a0aa9bf-bd59-4289-a86f-2e939cca0086 - [name: "test Project"] - removed
eu-west-1 - IoTSiteWiseAssetModel - 41dba598-ec81-465f-b96a-bad704924fe5 - [name: "Oven_Model", status: "ACTIVE"] - removed
eu-west-1 - IoTSiteWiseAssetModel - 74bd3cd0-5798-4b3a-930c-b0a1e3469492 - [name: "Mixer_Model", status: "ACTIVE"] - removed
eu-west-1 - IoTSiteWiseAssetModel - a3f6ff1f-4c29-4f45-8c80-b7a2f8b7f236 - [name: "testMaker_Model", status: "ACTIVE"] - removed
eu-west-1 - IoTSiteWiseAssetModel - b5ea342a-260a-4cb3-9d21-b4a2c4198b8f - [name: "HorizontalConveyor_Model", status: "ACTIVE"] - removed
eu-west-1 - IoTSiteWiseAssetModel - ba064ef9-1b60-44ed-a5e6-efb98e22172a - [name: "VerticalConveyor_Model", status: "ACTIVE"] - removed

Removal requested: 0 waiting, 0 failed, 0 skipped, 14 finished
```

ekristen commented on resources/iotsitewise-access-policy.go:
```diff
@@ -0,0 +1,150 @@
+	ID  *string
+}
+
+func (f *IoTSiteWiseAccessPolicy) Properties() types.Properties {
```
Rename to `r *IoTSiteWiseAccessPolicy` - reason standardization on `r` for pointer for all resource methods.

npellegrin replied:
Done in https://github.com/ekristen/aws-nuke/compare/94d209684ac69c83b3fde33a88e5a4aa4ef122da..99e1cf11e214fb625b141b80d4bee5158ee29915

ekristen commented on resources/iotsitewise-access-policy.go:
```diff
@@ -0,0 +1,150 @@
+}
+
+func (f *IoTSiteWiseAccessPolicy) Properties() types.Properties {
+	properties := types.NewProperties()
```
Replace with `return types.NewPropertiesFromStruct(r)` - reason standardizing on auto property detection based on struct, working on automated documentation generation a based on this data as well as filter and config validation

npellegrin replied:
Will it also work when one of the properties is a map, like resources with tags?

npellegrin replied:
Done in https://github.com/ekristen/aws-nuke/compare/94d209684ac69c83b3fde33a88e5a4aa4ef122da..99e1cf11e214fb625b141b80d4bee5158ee29915

ekristen reviewed (changes requested):
@npellegrin thanks for the PR. I'm working on standardizing the properties so that I can automatically generate documentation. Part of that change is to use exported properties on the structs in golang and use a new function `types.NewPropertiesFromStruct(r)`.

1. please rename all method pointers to `r`, I've made note on one of them so you can see.
2. change all properties to capital word case, I've made note on one of them so you can see.
3. change the properties to `return types.NewPropertiesFromStruct(r)`

ekristen commented:
If you'd prefer I can push a commit to address these things. I'm happy to assist. I always prefer to give folks to the option. I appreciate you opening a PR for all this.

npellegrin commented:
[x] renamed all method pointers to r
[x] changed all properties to capital word case
[x] changed the properties to return types.NewPropertiesFromStruct(r)
[x] Fixed a problem with asset hierarchy deletion

ekristen reviewed (approved):
Yes there is special provisions for maps and tags ;-)
